<!DOCTYPE html> <html lang="en">
<head>
	<title>MoDMaps3D - Main Menu</title>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css">
	<script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script> 
	<script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js"></script>
	<!-- <link rel="stylesheet" href="js/bootstrap.min.css">
	<script src="js/jquery.min.js"></script>
	<script src="js/bootstrap.min.js"></script> -->

	<script type="text/javascript" src="js/numeric-1.2.6.min.js"></script>
	<script type="text/javascript" src="js/jszip.min.js"></script>
	<script type="text/javascript" src="js/FileSaver.min.js"></script>
	<script type="text/javascript" src="js/parallel.js"></script>


	<script>
	var dbg = false, errorHasOccured = false, noMDS = false, splitBy, offlineCompMinSize = 500;
	var eigenval, eigenvec, eigensystem, finalEigenval, finalEigenvec, points, mapfile ;
	var fastaFilesLoaded = [], defaultInputType = "ncbi";
	var mapID, allSequences={}, distMatrix=[], fcgrRes=9, getTaxa = false;
	var accIDs, step1A, step1B, step2A, step2B, step3A, step3B, time;
	var callsNCBIDone, progress;
	var aidDistInfo = [], map_settings, previousdata = [];
	var labelColors = ['blue', 'red', 'green', 'orange', 'magenta', 'yellow', 'brown', 'lime', 'gray', 'pink','cyan','seagreen', 'olive', 'purple', 'black'];
	var tmpMap = false;

	function geturlparamvalue(name) {
		name = name.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]");
		var regexS = "[\\?&]" + name + "=([^&#]*)";
		var regex = new RegExp(regexS);
		var results = regex.exec(window.location.href);
		if(results!=null){
			return results[1];
		}else{
			return -1;
		}
	}

	function computeMDS(){
		time = new Date();
		step3A = time.getTime();
		console.log("BEGIN computeMDS()");	
		
		// pre-processign MDS

		// part 1
		var numOfSeq = distMatrix.length;
		var distMatrixSqTotals = [];
		for(var i=0; i<numOfSeq; i++){
			var tmpColSum = 0;
			for(var j=0; j<numOfSeq; j++){
				tmpColSum += Math.pow(distMatrix[i][j], 2);
			}
			distMatrixSqTotals.push(tmpColSum);
		}
		if(dbg){console.log('sqtotals=',distMatrixSqTotals);}
		
		// part 2
		var sumOfSqTotals = 0;
		for(var i=0; i<numOfSeq; i++){
			sumOfSqTotals += distMatrixSqTotals[i];
		}
		if(dbg){console.log('sumofsqtotals=',sumOfSqTotals);}
		
		// part 3
		var distMatrixCentered = [];
		for(var i=0; i<numOfSeq; i++){
			distMatrixCentered.push([]);
			for(var j=0; j<numOfSeq; j++){
				var tmpVal = -(1/2)*(Math.pow(distMatrix[i][j], 2) - (1/numOfSeq)*distMatrixSqTotals[i] - (1/numOfSeq)*distMatrixSqTotals[j] + (1/(Math.pow(numOfSeq, 2)))* sumOfSqTotals);
				distMatrixCentered[i].push(tmpVal);
			}
		}
		
		//DEBUGING FOR PRINTING dMatrix
		if(noMDS){
			if(dbg){console.log(distMatrixCentered);}
			
			$('#progress').html('<p>Please continue computation offline. <br> <strong>Instructions:</strong>\
				<ul><li>Download zip file (It is downloaded automatically)</li>\
				<li>Extract zip contents to a directory of your choice</li>\
				<li>Download either <a href="./extra/mds.nb" target="_blank">MDS-Mathematica</a> or <a href="./extra/mds.py" target="_blank">MDS-Python</a> code and place it in the same directory</li>\
				<li>Run the code, you should get an output file [map_output.txt]</li>\
				<li>Open [map_output.txt], copy its content and paste it <a href="./extra/input.html" target="_blank">here</a></li>\
				</ul></p>');
			console.log('COPY/PASTE MATHEMATICA');
			
			var toprint="{";
			for(var i=0; i<distMatrixCentered.length; i++){
				toprint+='{';
				for (var j=0; j<distMatrixCentered.length; j++){
					toprint+=distMatrixCentered[i][j].toFixed(10)+',';
				}
				toprint = toprint.substring(0, toprint.length - 1);
				toprint+='},'
			}
			toprint = toprint.substring(0, toprint.length - 1);
			toprint+='}';
			
			var toprint2='{';
			for(var i=0; i<accIDsSets.length; i++){
				toprint2 += '{';
				for(var j=0; j<accIDsSets[i].length; j++){
					toprint2 += '"'+accIDsSets[i][j]+'",';
				}
				toprint2 = toprint2.substring(0, toprint2.length - 1);
				toprint2 +='},'
			}
			toprint2 = toprint2.substring(0, toprint2.length - 1);
			toprint2 +='}';
			
			var toprint3='{';
			var totalNumOfSeq = Object.keys(allSequences).length;
			for(var i=0; i<totalNumOfSeq; i++){
				toprint3 += '{';
				toprint3 += '"'+allSequences[i][0]+'",';
				toprint3 += '"'+allSequences[i][1]+'",';
				toprint3 += '"'+allSequences[i][2]+'",';
				toprint3 += '"'+allSequences[i][4]+'",';
				toprint3 = toprint3.substring(0, toprint3.length - 1);
				toprint3 +='},'
			}
			toprint3 = toprint3.substring(0, toprint3.length - 1);
			toprint3 +='}';

			var toprint4 = '';
			for(var i=0; i<accIDsSets.length; i++){	toprint4 += accIDsSets[i].length + ",";	}
			toprint4 = toprint4.slice(0,-1);
			toprint4 += "\n";
			for(var i=0; i<accIDsSets.length; i++){ toprint4 += $("#colorset"+i).val() + ","; }
			toprint4 = toprint4.slice(0,-1);
			toprint4 += "\n";
			toprint4 += "5\nIndex,Acc,Name,Length,Taxa\n";
			for(var i=0; i<accIDsSets.length; i++){ toprint4 += $("#colorset"+i).val() + ","; }
			toprint4 = toprint4.slice(0,-1);
			toprint4 += "\n";
			for(var i=0; i<accIDsSets.length; i++){toprint4 +=$("#nameset"+i).val()+" ("+accIDsSets[i].length+"),";}
			toprint4 = toprint4.slice(0,-1);
			toprint4 += "\n";

			for(var i=0; i<$("#numofsets").val(); i++){ toprint4 += $("#nameset"+i).val() + ', '; }
			toprint4 = toprint4.slice(0,-2);  //because of space above!
			// OLD: toprint4 += '. Number of sequences is '+accIDs.length+ "\n";
			toprint4 += '#User-input sequences#'+accIDs.length+"#Approx.Inf.Dist#NA\n";
			
			
			// completely debugging part
			var toprint5="{";
			for(var i=0; i<distMatrix.length; i++){
				toprint5+='{';
				for (var j=0; j<distMatrix.length; j++){
					toprint5+=distMatrix[i][j].toFixed(10)+',';
				}
				toprint5 = toprint5.substring(0, toprint5.length - 1);
				toprint5+='},'
			}
			toprint5 = toprint5.substring(0, toprint5.length - 1);
			toprint5+='}';

			var zip = new JSZip();
			zip.file("dMatrix.txt", toprint);
			zip.file("accIDs.txt", toprint2);
			zip.file("allSequences.txt", toprint3);
			zip.file("mapheader.txt", toprint4);
			zip.file("debug.txt", toprint5);
			var content = zip.generate({type:"blob"});
			// see FileSaver.js
			saveAs(content, "dataForMMA.zip");
			return;
		}
		

		// EIGENVALUES - EIGENVECTORS
		try{
			res2 = numeric.eig(distMatrixCentered);	
		}catch(err){
			console.log(err.message);
			errorHasOccured = true;
			$('#progress').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span>'+' ERROR [<strong>'+err.message+'</strong>] while computing MDS. Internal errors originate from numerical errors in distance matrices. Consider changing/increasing value of length of k-mers.</div>');
			return;
		}
		if(dbg){console.log('numeric.js=',res2['lambda']['x']);}

		eigenval = res2['lambda']['x'];
		eigenvec = numeric.transpose(res2['E']['x']);   /// here transpose!!!
		eigensystem =[];
		for(var i=0; i<numOfSeq; i++){
			eigensystem.push([eigenval[i], eigenvec[i]]);
		}
		if(dbg){console.log("eigensystem=",eigensystem);}
		eigensystem.sort(function (a,b){ return Math.abs(b[0]) - Math.abs(a[0]) });
		if(dbg){console.log("eigensystem=",eigensystem);}
		

		// PICK 5 LARGEST AND POSITIVE
		finalEigenvec=[];
		finalEigenval=[];
		for(var i=0; i<numOfSeq; i++){
			if(eigensystem[i][0]>0){
				finalEigenvec.push(eigensystem[i][1]);
				finalEigenval.push(eigensystem[i][0]);
			}
			if(dbg){console.log("i-th eigeinvalue= [",i,"] [",eigensystem[i][0],"]");}
			if(finalEigenvec.length === 5){break;}
		}
		if(dbg){console.log("finalEigenval=",finalEigenval);}
		if(dbg){console.log("finalEigenvec=",finalEigenvec);}
		
		// POINTS COORDINATES
		points = numeric.dot(numeric.transpose(finalEigenvec), numeric.sqrt(numeric.T.diag(finalEigenval)['x']));
		if(dbg){console.log("points=",points);}
		
		// MIN-MAX PER DIMENSION (FOR LATER SCALING)
		var allMinMax =[], tmpMinMax = numeric.transpose(points);
		for(var i=0; i<5; i++){
			allMinMax.push( [ Math.min.apply(Math,tmpMinMax[i]) , Math.max.apply(Math,tmpMinMax[i]) ] );
		}
		if(dbg){console.log(allMinMax);}
		

		// PRODUCING MAPFILE
		mapfile = '';
		// line 1
		for(var i=0; i<accIDsSets.length; i++){
			mapfile += accIDsSets[i].length + ",";
		}
		mapfile = mapfile.slice(0,-1);
		mapfile += "\n";
		// line 2
		for(var i=0; i<accIDsSets.length; i++){
			mapfile += $("#colorset"+i).val() + ",";
		}
		mapfile = mapfile.slice(0,-1);
		mapfile += "\n";
		// line 3+4
		mapfile += "5\nIndex,Acc,Name,Length,Taxa\n";
		// line 5
		for(var i=0; i<accIDsSets.length; i++){
			mapfile += $("#colorset"+i).val() + ",";
		}
		mapfile = mapfile.slice(0,-1);
		mapfile += "\n";
		// line 6
		for(var i=0; i<accIDsSets.length; i++){
			mapfile += $("#nameset"+i).val()+" ("+accIDsSets[i].length+"),";
		}
		mapfile = mapfile.slice(0,-1);
		mapfile += "\n";
		
		// line 7
		// OLD mapfile += "No Description given\n";
		for(var i=0; i<$("#numofsets").val(); i++){
			mapfile += $("#nameset"+i).val() + ', ';
		}
		mapfile = mapfile.slice(0,-2);  //because of space above!
		// old mapfile
		// mapfile += '. Number of sequences is '+accIDs.length+ "\n";
		mapfile += '#User-input sequences#'+accIDs.length+"#Approx.Inf.Dist#NA\n";

		// rest of the lines
		for(var i=0; i<numOfSeq; i++){
			for(var j=0; j<points[i].length; j++){
				mapfile += ((2*points[i][j] - allMinMax[j][0] - allMinMax[j][1]) / ( allMinMax[j][1] - allMinMax[j][0] ) ) +"\n";	
			}
			mapfile += i+"\n";                      // INDEX
			mapfile += allSequences[i][0]+"\n";     // ACC
			mapfile += allSequences[i][4].split("|").pop().trim()+"\n";     // NAME
			mapfile += allSequences[i][1]+"\n";     // LENGTH
			mapfile += allSequences[i][2]+"\n";     // TAXA
		}
		mapfile = mapfile.slice(0,-1);
		if(dbg){console.log("c/p=\n",mapfile);}

		var mapTimeStamp = new Date().getTime();
		if(typeof(Storage) !== "undefined") {
			try{
				localStorage.setItem("mapfile"+mapTimeStamp, mapfile);
				localStorage.setItem("distMatrix"+mapTimeStamp, distMatrix);	
			}catch(err){
				console.log(err.message);
				errorHasOccured = true;
				$('#progress').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span>'+' ERROR [<strong>'+err.message+'</strong>]. Consider emptying localStorage.</div>');
				return;
			}
		} else {
			console.log("ERROR! NO LOCAL STORAGE AVAILABLE!?");
			alert("ERROR! NO LOCAL STORAGE AVAILABLE!?");
		}
		
		$('#progress').html('100%');
		$("#file_contents").html('<pre>'+mapfile+'</pre>');
		$("#file_contents").hide();
		// END OF MDS
		

		time = new Date();
		step3B = time.getTime();
		$("#stepstatus").html('\
			<p><strong>Step 1 (NCBI + FCGRs) DONE in ~'+(step1B-step1A)/1000+' sec</p>\
			<p>Step 2 (AID) DONE in ~'+(step2B-step2A)/1000+' sec</p>\
			<p>Step 3 (MDS) DONE in ~'+(step3B-step3A)/1000+' sec</p></strong>');
		$("#stepstatus").html($("#stepstatus").html()+'<p><a href="load.html?mapid=local'+mapTimeStamp+'" target="_blank">Show MoDMap</a>');
		if(dbg){
			$("#stepstatus").html($("#stepstatus").html()+' or <a href="#viewfile" id="viewfile">view contents</a>');
		}
		$("#stepstatus").html($("#stepstatus").html()+'</p>');
		$("#mainprogressbar").hide();
		$("#viewfile").click(function(){
			$("#file_contents").toggle();
		});

		// list local storage for debuging
		if(dbg){for (var a in localStorage) {console.log(a);}}
		if(dbg){console.log(Object.keys(localStorage));}
	}

	function computeAIDdist(){
		console.log("aid start");
		$("#aidinfo").empty();
		$("#aidinfomenu").slideDown();
		$("#aidinfo").html('<img src="img/loading.gif" width="30" height="30"/>');
		console.log($("#seq0").val());
		console.log($("#seq1").val());
		errorHasOccured = false;

		var gotBothSequences = 0;
		for (var indSeq=0; indSeq<2; indSeq++){
			loadFastaFromNCBI(indSeq, $("#seq"+indSeq).val(), function(ind, accID, output, header){
				aidDistInfo[ind] = [accID, header, output.length,  buildFCGR(output, fcgrRes)];
				if(dbg){console.log([ind,accID,output.length, header,'DONE']);}
				
				gotBothSequences += 1;
				if(gotBothSequences == 2 && !errorHasOccured){
					
					console.log("finished computation");
					var aidinfotext = "<strong>NCBI: ["+aidDistInfo[0][0]+"]</strong> - Length: "+aidDistInfo[0][2]+"<br>Header: "+aidDistInfo[0][1]+"<br><br> <strong>NCBI: ["+aidDistInfo[1][0]+"]</strong> - Length: "+aidDistInfo[1][2]+"<br>Header: "+aidDistInfo[1][1]+"<br><br>";
					
					var aidnumerator = aidDistInfo[0][3]['size'] + aidDistInfo[1][3]['size'];
					var aiddenominator=0;
					for(var unionInd=0; unionInd<Math.pow(4,fcgrRes); unionInd++){
						if( (aidDistInfo[0][3]['flatFCGR'][unionInd]==1) || (aidDistInfo[1][3]['flatFCGR'][unionInd]==1) ){
							aiddenominator++;
						} 
					}
					console.log(aidnumerator,aiddenominator,2 - aidnumerator*1.0/aiddenominator);
					
					aidinfotext += "<strong>AID: </strong>"+ String((2 - aidnumerator*1.0/aiddenominator).toFixed(5));
					$("#aidinfo").html(aidinfotext);


				}

			});	
		}

		console.log("aid end");
	}

	function computeDistMatrix(){
		time = new Date();
		step2A = time.getTime();
		console.log("BEGIN computeDistMatrix()");	
		$("#stepstatus").html('\
			<p><strong>Step 1 (NCBI + FCGRs) DONE in ~'+(step1B-step1A)/1000+' sec</strong></p>\
			<p>Step 2 of 3: (AID)</p>');
		$('#progress').html('0%');

		var dim=accIDs.length;
		distMatrix=[];
		for(var i=0; i<dim; i++){
			distMatrix.push([]);
			for(var j=0; j<dim; j++){
				distMatrix[i].push(0);
			}
		}
		
		splitBy = Math.max(Math.ceil(dim/5), 50);
		console.log("splitBy= ",splitBy);
		var input = []
		for(var i=1; i<= Math.ceil(dim/splitBy); i++){
			for(var j=i; j<= Math.ceil(dim/splitBy); j++){
				var minX, maxX, minY, maxY;
				minX = (i-1)*splitBy;
				maxX = Math.min(i*splitBy-1,dim-1);
				minY = (j-1)*splitBy;
				maxY = Math.min(j*splitBy-1,dim-1);
				var allSequencesX = [] , allSequencesY = [];
				for(var i00=minX; i00<=maxX; i00++){
					allSequencesX.push(allSequences[i00]);
				}
				for(var i00=minY; i00<=maxY; i00++){
					allSequencesY.push(allSequences[i00]);
				}
				input.push([minX, maxX, minY, maxY, allSequencesX, allSequencesY, fcgrRes]);
			}
		}
		console.log("input= ",input);
		$('#progress').html('Parallel computation of '+input.length+' matrices of ~ '+splitBy+'x'+splitBy+' each. Please wait..');
		var p = new Parallel( input );
		
		var computeChunks = function (chunk) {
			var bgX = chunk[0], endX = chunk[1], bgY = chunk[2], endY = chunk[3];
			var allSequencesX = chunk[4], allSequencesY = chunk[5], fcgrRes = chunk[6];
			var res=[], tmpRow, numerator, denominator ;
			var timeBegin = new Date();
			console.log("Worker ["+bgX+","+endX+" - "+bgY+","+endY+"]");

			for(var i = 0; i<allSequencesX.length; i++){
				tmpRow=[];
				for(var j = 0; j<allSequencesY.length; j++){
					numerator = allSequencesX[i][3]['size'] + allSequencesY[j][3]['size'];
					denominator=0;
					for(var unionInd=0; unionInd<Math.pow(4,fcgrRes); unionInd++){
						if( (allSequencesX[i][3]['flatFCGR'][unionInd]==1) || (allSequencesY[j][3]['flatFCGR'][unionInd]==1) ){
							denominator++;
						} 
					}
					// console.log(i,j,'=',numerator,denominator,2 - numerator*1.0/denominator);
					tmpRow.push(2 - numerator*1.0/denominator);
				}
				res.push(tmpRow);
			}

			var timeEnd = new Date();
			console.log("Worker ["+bgX+","+endX+" - "+bgY+","+endY+"] execution time: "+String((timeEnd-timeBegin)/1000)+" sec");
			return [bgX, endX, bgY, endY, res];
		};
 

		function assembleChunks(data){
			// console.log("data=",data,data.length);
			// console.log("dMat=",distMatrix.length);

			for(var i=0; i<data.length; i++){
				var tmpInfo = data[i];
				// console.log("tmpInfo=",i,tmpInfo);
				var bgX = tmpInfo[0], bgY = tmpInfo[2];
				var endX = tmpInfo[1], endY = tmpInfo[3];
				for(var indexRow=0; indexRow<= endX - bgX; indexRow++){
					for(var indexCol=0; indexCol<= endY -bgY; indexCol++){
						// console.log(bgX + indexRow, bgY + indexCol, tmpInfo[4][indexRow][indexCol],distMatrix[bgX + indexRow][bgY + indexCol]);
						distMatrix[bgX + indexRow][bgY + indexCol] = tmpInfo[4][indexRow][indexCol];
						distMatrix[bgY + indexCol][bgX + indexRow] = tmpInfo[4][indexRow][indexCol];
						
					}
				}
			}	
			
			// if(dbg){console.log(distMatrix);}
			// var toprint="{";
			// for(var i=0; i<distMatrix.length; i++){
			// 	toprint+='{';
			// 	for (var j=0; j<distMatrix.length; j++){
			// 		toprint+=distMatrix[i][j]+',';
			// 	}
			// 	toprint = toprint.substring(0, toprint.length - 1);
			// 	toprint+='},'
			// }
			// toprint = toprint.substring(0, toprint.length - 1);
			// toprint+='}';
			// console.log('COPY/PASTE MATHEMATICA');
			// console.log("Eigensystem[",toprint,"]");


			time = new Date();
			step2B = time.getTime();		
			$("#stepstatus").html('\
				<p><strong>Step 1 (NCBI + FCGRs) DONE in ~'+(step1B-step1A)/1000+' sec</p>\
				<p>Step 2 (AID) DONE in ~'+(step2B-step2A)/1000+' sec</p></strong>\
				<p>Step 3 of 3: (MDS)</p>');
			$('#progress').html('Please wait.. (browser may freeze for a while) <img src="img/loading.gif" width="30" height="30"/>');
			
			setTimeout(computeMDS, 500); 

		}

		p.map(computeChunks).then(assembleChunks);		
	}

	function buildFCGR(seq, k){
		var numOfKmers=0, curmer, curmerInd, kmers={}, mapACGT={"A":0, "C":1, "G":2, "T":3};
		var newseq = seq.replace(/B|D|E|F|H|I|J|K|L|M|N|O|P|Q|R|S|U|V|W|X|Y|Z/gi, "");
		var kmersInds = new Uint8Array(Math.pow(4, k));
		if(dbg){console.log((seq.length - newseq.length)+' lost bp');}
		for(var i=0; i<newseq.length-k+1; i++){
			curmer = newseq.slice(i,i+k);
			curmerInd = 0;
			for(var j=0; j<curmer.length; j++){ 
				curmerInd += Math.pow(4, k-1-j)*mapACGT[curmer[j]]; 
			}
			
			if(kmersInds[curmerInd]==0){
				numOfKmers++;
				kmersInds[curmerInd]=1;
			}

		}
		if(dbg){console.log(numOfKmers+' #kmers '+kmersInds.length);}
		return {"size":numOfKmers, "flatFCGR":kmersInds};
	}

	function loadFastaFromNCBI(ind, accID, callback){
		$.ajax({
			url: "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&rettype=fasta&retmode=text&id="+accID,
			success: function(result){
				var data = result.split(/\n\r?/gi);
				var header;
				if( data.length && data[0][0]==='>' ){ header = data[0]; }
				while (data.length && data[0][0] === '>') {data.shift();}
				var outputFasta=data.join('');
				if(dbg){console.log('['+accID+'] len='+outputFasta.length);}
				callback(ind, accID, outputFasta, header);
			},
			error: function(xhr, status, error) {
				console.log('STATUS=',status,'ERROR from loadFastaFromNCBI=',error);
				errorHasOccured = true;
				$('#progress').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span>'+' ERROR [<strong>'+error+'</strong>] while fetching data from NCBI for [<strong>'+accID+'</strong>]</div>');
				$('#aidinfo').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span>'+' ERROR [<strong>'+error+'</strong>] while fetching data from NCBI for [<strong>'+accID+'</strong>]</div>');

			}
		});
	}

	function getTaxaFromNCBI(ind, idNCBI){
		if(dbg){console.log("getting taxa for id=",ind);}
		$.ajax({
			url: "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=gb&retmode=xml&id="+idNCBI,
			success: function(result){
				// res=result;
				// console.log("res=", result);
				var taxaString = new XMLSerializer().serializeToString(result);
				var taxa = /<GBSeq_taxonomy>(.*)<\/GBSeq_taxonomy>/g.exec(taxaString);
				if(dbg){console.log("taxa=", taxa[1]);}
				allSequences[ind][2] = taxa[1].trim();
				
				callsNCBIDone += 0.5;
				progress = callsNCBIDone*100.0/accIDs.length;
				progress = progress.toFixed(2);
				if(!errorHasOccured){$('#progress').html(progress+'%');}
				if(callsNCBIDone == accIDs.length && !errorHasOccured){
					time = new Date();
					step1B = time.getTime();
					computeDistMatrix();
				}
			},
			error: function(xhr, status, error) {
				console.log('STATUS=',status,'ERROR getTaxaFromNCBI=',error);
				errorHasOccured = true;
				$('#progress').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span>'+' ERROR [<strong>'+error+'</strong>] while fetching Taxonomy data from NCBI for [<strong>'+idNCBI+'</strong>]</div>');
			}
		});		
	}

	$(document).ready(function(){

		if(geturlparamvalue("tmpmaps")=="true"){
			console.log("Looking at tmp directory NOW..");
			tmpMap = true;
		}

		$.ajax({
			url: "./extra/maps.txt", 
			cache: false, 
			success: function(result){
			// LOAD FILENAMES OF MAPS 
			//var result = JSON.parse(result);
			//console.log(result);
			var resfetch;
			if(tmpMap){
				resfetch=JSON.parse(result)['tmpmaps'];
			}else{
				resfetch=JSON.parse(result)['maps'];	
			}
			var outcome='';
			for(i=0;i<resfetch.length;i++){
				outcome+='<div class="panel-heading"><h4 class="panel-title"><a data-toggle="collapse" data-parent="#modmap-accordion-'+i+'" href="#modmap-collapse-'+i+'">'+resfetch[i][0]+'&nbsp;&nbsp;<span class="badge">'+resfetch[i][1].length+'</span></a></h4></div>';
				var tmp='<div id="modmap-collapse-'+i+'" class="panel-collapse collapse out"><div class="panel-body"><ul class="list-group large-group">';
				for(var j=0; j<resfetch[i][1].length; j++){
					tmp+='<li class="list-group-item"><a href="#">'+resfetch[i][1][j]+'</a></li>';
				}
				tmp+='</ul></div></div>';
				outcome+=tmp;
			}
			$('#listgroupsofmaps').html(outcome);

			// LOAD PREVIEW CONTENT FOR EACH FILENAME
			$(".list-group-item").click(function() {
				$("#filecontent").html('<img src="img/loading.gif" width="50" height="50"/>');
				console.log('Loading = '+$(this).text());
				mapID=$(this).text();
				
				var mapPathToLoad = "";
				var mapPathToLoadInURL = "";
				if(tmpMap){
					mapPathToLoad = "./maps/tmp/"+mapID;
					mapPathToLoadInURL = "tmp/"+mapID;
				}else{
					mapPathToLoad = "./maps/"+mapID;
					mapPathToLoadInURL = mapID;
				}

				$.ajax({
					url: mapPathToLoad, 
					cache: false,
					success: function(result){
						var resfetch=result.split("\n");

						if(resfetch[6].indexOf("#") > -1){
							// NEW TYPE OF LEGEND WITH INFO SPLIIT BY #
							console.log("Found legend, split by #");
							var sublegends = resfetch[6].split("#");
							var subleg = "<div>&bull; "+sublegends[0]+"</div>";
							subleg += "<div>&bull; "+sublegends[1]+"</div>";
							subleg += "<div>&bull; "+sublegends[2]+"</div>";
							subleg += "<div>&bull; "+sublegends[3]+"</div>";
							subleg += "<div>&bull; Avg.Seq.Length: "+sublegends[4]+"</div>";
							
							$("#filecontent").html('<p id="mapId"><strong>'+resfetch[5]+'</strong></p><p>'+subleg+'</p><a href="load.html?mapid='+mapPathToLoadInURL+'" target="_blank">Show Map</a><p></p>');
						}else{
							// BACKWARD COMPATIBILITY LEGEND
							console.log("BACKWARD COMPATIBILITY LEGEND");
							$("#filecontent").html('<p id="mapId"><strong>'+resfetch[5]+'</strong></p><p>'+resfetch[6]+'</p><a href="load.html?mapid='+mapPathToLoadInURL+'" target="_blank">Show Map</a><p></p>');
						}

						
					}
				});
			});
		}});

		$("#goto1").click(function(){$('#guide-tabs li:eq(1) a').tab('show');});
		$("#goto2").click(function(){$('#guide-tabs li:eq(2) a').tab('show');});
		$("#goto3").click(function(){$('#guide-tabs li:eq(3) a').tab('show');});
		$("#kmerslen").val('9');
		$("#getTaxa").val('0');
		
		// $('#guide-tabs li:eq(2) a').tab('show');

		$("#buildmapmenucontent").slideUp();
		$("#extendmapmenucontent").slideUp();
		$("#inputdatamenu").slideUp();
		$("#computemenu").slideUp();
		$("#hotkeyslist").slideUp();
		
		$("#hotkeysshowhide").click(function(){
			$("#hotkeyslist").slideToggle();
		});

		$("#buildmapmenuheader").click(function(){
			// console.log("build");
			$("#buildmapmenucontent").slideDown();
			$("#inputdatamenu").slideDown();
			$("#extendmapmenucontent").slideUp();
			$("#computemenu").slideUp();
			$("#numofsets").val(0);
			$("#numofsets").trigger("change");
			// $("#selectnumofsets").css({"border-color": "red", "border-width":"5px", "border-style":"solid"});
		});

		$("#extendmapmenuheader").click(function(){
			// console.log("extend");
			$("#buildmapmenucontent").slideUp();
			$("#inputdatamenu").slideUp();
			$("#extendmapmenucontent").slideDown();
			$("#computemenu").slideUp();
		});


		$("#aid1").click(function(){
			$("#seq0").empty();
			$("#seq0").val("NC_012920");
			$("#seq1").empty();
			$("#seq1").val("NC_001643");	
			computeAIDdist();		
		}); 
		$("#aid2").click(function(){
			$("#seq0").empty();
			$("#seq0").val("NC_012920");
			$("#seq1").empty();
			$("#seq1").val("NC_011137");
			computeAIDdist();					
		});  

		$("#dataset1").click(function(){
			$.ajax({
				url: "./maps/samplemaps/sample1.txt", 
				cache: false,
				success: function(result){
				var result = JSON.parse(result);
				console.log("Loading: ",result["mapDescription"]);
				$('#allInputData').slideDown();
				$("#file_contents").empty();
				$("#file_contents").hide();
				$("#numofsets").val(result["numofsets"]);
				$("#numofsets").trigger("change");
				for(var i=0; i<result["namesets"].length; i++){
					$("#set"+i).html(result["sets"][i]);
					$("#nameset"+i).val(result["namesets"][i]);
					$("#colorset"+i).val(result["colorsets"][i]);		
				}
				$("#stepstatus").hide();
				$("#mainprogressbar").hide();
				$("#computemenu").slideUp();
			}});		
		});  
		$("#dataset2").click(function(){
			$.ajax({
				url: "./maps/samplemaps/sample2.txt", 
				cache: false,
				success: function(result){
				var result = JSON.parse(result);
				console.log("Loading: ",result["mapDescription"]);
				$('#allInputData').slideDown();
				$("#file_contents").empty();
				$("#file_contents").hide();
				$("#numofsets").val(result["numofsets"]);
				$("#numofsets").trigger("change");
				for(var i=0; i<result["namesets"].length; i++){
					$("#set"+i).html(result["sets"][i]);
					$("#nameset"+i).val(result["namesets"][i]);
					$("#colorset"+i).val(result["colorsets"][i]);		
				}
				$("#stepstatus").hide();
				$("#mainprogressbar").hide();
				$("#computemenu").slideUp();
			}});
		});
		$("#dataset3").click(function(){
			$.ajax({
				url: "./maps/samplemaps/sample3.txt", 
				cache: false,
				success: function(result){
				var result = JSON.parse(result);
				console.log("Loading: ",result["mapDescription"]);
				$('#allInputData').slideDown();
				$("#file_contents").empty();
				$("#file_contents").hide();
				$("#numofsets").val(result["numofsets"]);
				$("#numofsets").trigger("change");
				for(var i=0; i<result["namesets"].length; i++){
					$("#set"+i).html(result["sets"][i]);
					$("#nameset"+i).val(result["namesets"][i]);
					$("#colorset"+i).val(result["colorsets"][i]);		
				}
				$("#stepstatus").hide();
				$("#mainprogressbar").hide();
				$("#computemenu").slideUp();
			}});
		});

		$("#kmerslen").change(function(){
			fcgrRes = parseInt($("#kmerslen").val());
			console.log("NEW fcgrRes=",fcgrRes);
		});
		$("#getTaxa").change(function(){
			getTaxa = parseInt($("#getTaxa").val());
			console.log("NEW getTaxa=", getTaxa);
		});

		$("#computedist").click(computeAIDdist);

		$("#loadrefmap").click(function(){
			console.log("Loading RefMap: ", $("#extendmapid").val());
			$("#inputdatamenu").slideDown();

			$.ajax({
				url: "./maps/refmaps/refmap_"+$("#extendmapid").val()+".txt", 
				cache: false,
				success: function(result){
				console.log("./maps/refmaps/refmap_"+$("#extendmapid").val()+".txt");
				var result = JSON.parse(result);
				console.log("RefMap Description: ",result["mapDescription"]);
				
				$('#allInputData').slideDown();
				$("#file_contents").empty();
				$("#file_contents").hide();
				$("#numofsets").val(result["numofsets"]+1);
				$("#numofsets").trigger("change");

				for(var i=0; i<result["namesets"].length; i++){
					$("#set"+i).html(result["sets"][i]);
					$("#nameset"+i).val(result["namesets"][i]);
					$("#colorset"+i).val(result["colorsets"][i]);		
				}

				$("#nameset"+result["namesets"].length).val("NEW_points");
				$("#colorset"+result["namesets"].length).val("black");	
				if( defaultInputType == "ncbi"){
					$("#set"+result["namesets"].length).html($("#extrancbi").val());
				}
				if( defaultInputType == "fasta"){
					var tmpOut = [];
					for(var i=0; i<fastaFilesLoaded.length; i++){
						tmpOut.push("fastaFile"+String(i+1));
					}
					// console.log(tmpOut);
					$("#set"+result["namesets"].length).html(tmpOut.join(","));
				}
				

				$("#stepstatus").hide();
				$("#mainprogressbar").hide();
				$("#computemenu").slideUp();
			}});
		});

		$("#buildmap").click(function(){
			noMDS = false;
			$("#computemenu").slideDown();
			$('#allInputData').slideUp();
			$('#allInputDataMinimized').slideDown();
			console.log("localStorageKeys before=",Object.keys(localStorage));
			localStorage.clear();
			console.log("localStorageKeys after=",Object.keys(localStorage));
			time = new Date();
			step1A = time.getTime();
			$("#file_contents").empty();
			$("#file_contents").hide();
			$("#mainprogressbar").show();
			$("#stepstatus").show();
			$("#stepstatus").html('Step 1 of 3: (NCBI + FCGRs)');
			$('#progress').html('0%'); 
			errorHasOccured = false;
			
			// GET "accIDs"
			accIDsSets=[];
			for(var i=0; i<$("#numofsets").val(); i++){
				console.log(i,$("#set"+i).val().split(","));
				var tmpSet = $("#set"+i).val().split(",");
				if(tmpSet[0]!=""){
					accIDsSets.push(tmpSet);
				}
			}
			accIDs = [].concat.apply([], accIDsSets);
			console.log("accIDs=",accIDs);
			console.log("In total = " + accIDs.length + " sequences.");
			if(accIDs.length < 5){
				errorHasOccured = true;
				$('#progress').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span>'+' ERROR [<strong>Found '+accIDs.length+' sequences.</strong>] Please enter at least 5 sequences and try again.]</div>');
				return;
			}
			if(accIDs.length > offlineCompMinSize ){
				noMDS = true;
				alert("Your input contains more than ["+offlineCompMinSize+"] sequences. This requires that the last step (MDS) has to be computed offline using either Mathematica or Python. Files and code will be provided to you once you reach that point.");
			}

			// AJAX CALLS TO NCBI FOR FASTA FILES and/or READING USER FASTA FILES
			fastaFilesLoadedIndex = 0;
			callsNCBIDone = 0;
			for(var i=0; i<accIDs.length; i++){
				if(!errorHasOccured){

					if(accIDs[i].trim().substr(0,9) == "fastaFile"){
						
						// IF IT'S UPLOADED FASTA FILE
						var reader = new FileReader();
						reader.onload = (function(f,posInAllSeq, accInAllSeq) {
							return function(e) { 
								console.log("fasta: ", accInAllSeq, "fileInfo: ", e.target);
								var data = e.target.result.split(/\n\r?/gi);
								var header;
								if( data.length && data[0][0]==='>' ){ header = data[0]; }
								while (data.length && data[0][0] === '>') {data.shift();}
								var outputFasta=data.join('');

								allSequences[posInAllSeq]=[accInAllSeq, outputFasta.length, "NotAvailable", buildFCGR(outputFasta, fcgrRes), header];

								if(true){console.log([posInAllSeq,accInAllSeq,outputFasta.length,'DONE']);}
								callsNCBIDone += 1;

								// UPDATE PROGRESS AND START computeDistMatrix() IF LAST SEQUENCE IS FASTA
								progress = callsNCBIDone*100.0/accIDs.length;
								progress = progress.toFixed(2);
								if(!errorHasOccured){$('#progress').html(progress+'%');} 
								if(callsNCBIDone == accIDs.length && !errorHasOccured){
									time = new Date();
									step1B = time.getTime();
									computeDistMatrix();
								}
							}
						})(fastaFilesLoaded[fastaFilesLoadedIndex], i, accIDs[i].trim());
						
						reader.readAsText(fastaFilesLoaded[fastaFilesLoadedIndex]);
						fastaFilesLoadedIndex += 1;

					}else{

						// IF IT'S NCBI ACCESSION NUMBER
						loadFastaFromNCBI(i, accIDs[i].trim(), function(ind, accID, output, header){
							allSequences[ind]=[accID, output.length, /**/, buildFCGR(output, fcgrRes), header];
							if(getTaxa){
								getTaxaFromNCBI(ind, accID);
							}else{
								callsNCBIDone += 0.5;
								allSequences[ind][2] = "NotAvailable";
							}
							if(dbg){console.log([ind,accID,output.length,'DONE']);}
							callsNCBIDone += 0.5;

							// UPDATE PROGRESS AND START computeDistMatrix() IF LAST SEQUENCE IS NCBI
							progress = callsNCBIDone*100.0/accIDs.length;
							progress = progress.toFixed(2);
							if(!errorHasOccured){$('#progress').html(progress+'%');} 
							if(callsNCBIDone == accIDs.length && !errorHasOccured){
								time = new Date();
								step1B = time.getTime();
								computeDistMatrix();
							}
						});
					}

				}
			}
		});
		
		$("#allInputDataShow").click(function(){
			$("#allInputData").slideToggle();
		});

		$("#numofsets").change(function(){
			if($("#numofsets").val() == 0){
				$("#numofsets").css({"border-color": "red", "border-width":"5px", "border-style":"solid"});
			}else{
				// DEFAULT VALUES as reported in chrome by jquery
				// $("#numofsets").css("border-color");
				// $("#numofsets").css("border-width");
				// $("#numofsets").css("border-style");
				$("#numofsets").css({"border-color": "rgb(169, 169, 169)", "border-width":"0.666667px", "border-style":"solid"});
			}
			
			var colorList = '';
			for(var i=0; i<labelColors.length; i++){
				colorList += '<option value="'+labelColors[i]+'">'+ labelColors[i]+'</option>';
			}

			// save previous accessions used
			previousdata = [];
			for(var i=0; i<15; i++){
				if($("#set"+i).html() != undefined){
					previousdata.push([ $("#nameset"+i).val(), $("#colorset"+i).val(), $("#set"+i).html() ]);
				}
			}
			console.log("previous_accessions",previousdata);


			console.log("numofsets=",$("#numofsets").val());
			$("#sets").html('');
			for(var i=0; i<$("#numofsets").val(); i++){
				$("#sets").html($("#sets").html()+'<p>GroupName: <input type="text" size="5" id="nameset'+i+'" value="Set['+i+']"> GroupColor: <select id="colorset'+i+'">'+colorList+'</select><br>Insert comma separated NCBI accession numbers:<br><textarea id="set'+i+'" autofocus  rows="2" cols="40"></textarea></p>');
			}
			for(var i=0; i<$("#numofsets").val(); i++){$("#colorset"+i).val(labelColors[i]);}

			//restore saved accessions
			for(var i=0; i<previousdata.length; i++){
				if($("#set"+i).html() != undefined){
					$("#nameset"+i).val(previousdata[i][0]);
					$("#colorset"+i).val(previousdata[i][1]);
					$("#set"+i).html(previousdata[i][2]);
				}
			}


			if($("#numofsets").val()>0){$("#buildmap").show();}else{$("#buildmap").hide();}
		});

		var tmpDataset = geturlparamvalue('dataset');
		if(tmpDataset!=-1 && tmpDataset.slice(0,5)=='local'){
			newDataset = localStorage.getItem("dataset"+tmpDataset.slice(5));
			$("#buildmapmenuheader").click();
			$('#guide-tabs li:eq(2) a').tab('show');
			$("#file_contents").empty();
			$("#file_contents").hide();
			$("#numofsets").val('1');
			$("#numofsets").trigger("change");
			$("#set0").html(newDataset);
			$("#stepstatus").hide();
			$("#mainprogressbar").hide();
			
			// he came here for building, not extending, right?!
			// OR
			// $("#buildmapmenucontent").slideDown();
			// $("#inputdatamenu").slideDown();
			// $("#extendmapmenucontent").slideUp();
			// $("#computemenu").slideUp();


			//$("#buildmap").click(); allow user to change settings if wanted
		}else{
			//console.log('no local dataset');
		}

		$("#fastafiles").hide();
		$('input[type=radio][name=extrapointstype]').change(function() {
      if (this.value == 'ncbi') {
      	defaultInputType = "ncbi";
        console.log("ncbi selected..");
        $("#extrancbi").show();
      	$("#fastafiles").hide();
        $("#fastafilesinfo").hide();
        $("#inputdatamenu").slideUp();
      	$("#computemenu").slideUp(); 
      	$("#loadrefmap").show();   	
      }
      else if (this.value == 'fasta') {
      	defaultInputType = "fasta";
				console.log("fasta selected..");
				$("#extrancbi").hide();
      	$("#fastafiles").show();
				$("#fastafilesinfo").show();
        $("#inputdatamenu").slideUp();
      	$("#computemenu").slideUp();
      	$("#loadrefmap").hide();

      	// Check for the various File API support.
				if(window.File && window.FileReader && window.FileList && window.Blob){
					console.log("everything ok");
				}else{
					console.log('The File APIs are not fully supported in this browser.');
					$('#fastafiles').hide()
					$('#fastafilesinfo').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span> ERROR <strong>The File APIs are not fully supported in this browser.</strong> Please use another browser or input NCBI accession numbers instead.</div>');
					$("#inputdatamenu").slideUp();
      		$("#computemenu").slideUp(); 
				}

				if(fastaFilesLoaded.length >0){
		    	$("#loadrefmap").show();
		    } 
      }
    });
	
		$("#fastafiles").change(function(evt){
			console.log("Loading fasta files metadata..");
			var files = evt.target.files; 
			fastaFilesLoaded = files;
			var output = [];
	    for (var i = 0, f; f = files[i]; i++) {
	      output.push('<li><strong>', escape(f.name), '</strong> (', f.type || 'n/a', ') - ', f.size, ' bytes, last modified: ', f.lastModifiedDate ? f.lastModifiedDate.toLocaleDateString() : 'n/a','</li>');
	    }
	    $("#fastafilesinfo").html('<ul>' + output.join('') + '</ul>');
	    $("#loadrefmap").show();
		});

	
	});

	</script>

	<style>
	body {
		margin-bottom: 60px;
	}
	#listgroupsofmaps{
		margin: 10px;
	}
	#filecontent{
		margin: 10px;
	}
	</style>
</head>
<body>

	<div class = "page-header">
		<h1 class="text-center">MoDMaps3D</h1>
	</div>

	<div class="container">
		<ul class="nav nav-tabs" id="guide-tabs">
			<li class="active"><a data-toggle="tab" href="#home">Home</a></li>
			<li><a data-toggle="tab" href="#menu1">Explore a Map</a></li>
			<li><a data-toggle="tab" href="#menu2">Build or Extend a Map</a></li>
			<li><a data-toggle="tab" href="#menu3">Compute a distance</a></li>
			<li><a data-toggle="tab" href="#menu4">About</a></li>
		</ul>

		<div class="tab-content">
			<div id="home" class="tab-pane fade in active">
				<h3>Welcome to MoDMaps3D!</h3>
				<p>Here are some of the things you can do with three dimensional Molecular Distance Maps.</p>

				<dl class="dl-horizontal">
					<dt><a href="#" id="goto1">Explore a Map</a></dt>
					<dd>Explore available built-in MoDMaps. These maps provide extensive information for each genomic sequence involved in each map, along with complementary information.</dd>
					<dt><a href="#" id="goto2">Build or Extend a Map</a></dt>
					<dd>Build a new MoDMap from scratch tailored to your dataset. Include up to 12 categories, with as many sequences you need in each of them. Alternatively, extend a MoDMap adding your points. You can select a "base" map from a variety of available maps and then add your DNA sequences in it.</dd>
					<dt><a href="#" id="goto3">Compute a distance</a></dt>
					<dd>Compute the distance between any two DNA sequences, given as NCBI accession numbers.</dd>
				</dl>
				<div id="fastafile"></div>
			</div>
			<div id="menu1" class="tab-pane fade">
				<h3>Explore available MoDMaps</h3>
				<p>Here you can find a list of available built-in MoDMaps to explore. Select a map from sidebar, preview its content, and click on "Show Map" to see it in action.</p>
				
				<div class="col-xs-5">
					<div id="listgroupsofmaps" class="panel panel-default">
					</div>
				</div>
				<div class="col-xs-6 " id="filecontent">
				</div>  
			</div>
			<div id="menu2" class="tab-pane fade">
				<!-- style="padding-right:0px; border-right: 1px solid #ccc;" -->
				<div class="col-xs-10" id="buildmapview" >
					<h3>Build your map, for your dataset</h3>
					
					<!-- <div class="panel panel-default">
					  <div class="panel-heading">Panel with panel-default class</div>
					  <div class="panel-body">Panel Content</div>
					</div> -->

					<div class="panel panel-default">
						<div class="panel-heading"><label>Choose an option to continue</label></div>
						<div class="panel-body">
						<label><input type="radio" id="buildmapmenuheader" name="mapmenu"/> Build a MoDMap from your dataset</label><br>
						<label><input type="radio" id="extendmapmenuheader" name="mapmenu"/> Extend an existing MoDMap</label>
						</div>
					</div>

					<div class="panel panel-default" id="buildmapmenucontent">
						<div class="panel-heading"><label>Info</label></div>
						<div class="panel-body">
							<p>If you do not know the accession number of a species, you can search directly in NCBI from <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Homo+sapiens" target="_blank">here</a>. Samples you can try: 
							<a href="#" id="dataset1">[1] 62 mtDNA primates</a> - 
							<a href="#" id="dataset2">[2] 112 mtDNA amphibia</a> - 
							<a href="#" id="dataset3">[3] 39 HIV-1 genomes</a> 
							</p>
						</div>
					</div>

					<div class="panel panel-default" id="extendmapmenucontent">
						<div class="panel-heading"><label>Info</label></div>
						<div class="panel-body">
							<p>Extend a map by adding new points in a set of predefined MoDMaps. The new points can be given as comma separated accession numbers from NCBI. If you do not know the accession number of a species, you can search directly in NCBI from <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Homo+sapiens" target="_blank">here</a>.</p>


							<p>
							<dl class="dl-horizontal">
							<dt>Background map:</dt>
							<dd><select id="extendmapid">
								<option value="Animalia_mtDNA_ClassAmphibia">Animalia_mtDNA_ClassAmphibia</option>
								<!-- INSECTS REMOVED BECAUSE OF TOO MANY SUBSETS 21>>10 -->
								<!-- <option value="Animalia_mtDNA_ClassInsecta">Animalia_mtDNA_ClassInsecta</option> -->
								<option value="Animalia_mtDNA_ClassMammalia">Animalia_mtDNA_ClassMammalia</option>
								<option value="Animalia_mtDNA_OrderPrimates">Animalia_mtDNA_OrderPrimates</option>
								<option value="Animalia_mtDNA_Vertebrata">Animalia_mtDNA_Vertebrata</option>
								<option value="Animalia_mtDNA_Ins_Mam_Amph">Animalia_mtDNA_Ins_Mam_Amph</option>
								<option value="Fungi_mtDNA">Fungi_mtDNA</option>
								<option value="Plants_mtDNA">Plants_mtDNA</option>
								<option value="Protists_mtDNA">Protists_mtDNA</option>
								<option value="Protists_ptDNA">Protists_ptDNA</option>
								<option value="Archaea_nDNA">Archaea_nDNA</option>
								<option value="Bacteria_nDNA">Bacteria_nDNA</option>
								<option value="Viruses_HIV1">Viruses_HIV1</option>
								<option value="Haplogroups_All">Haplogroups_All</option>
							</select><br></dd>
							<dt>Extra Sequences:</dt>
							<dd>
								<label><input type="radio" name="extrapointstype" value="ncbi" checked="checked">
								Insert comma separated NCBI accession numbers</label><br>
								<label><input type="radio" name="extrapointstype" value="fasta">
								Load FASTA files</label><br>
								<textarea id="extrancbi" rows="2" cols="40"></textarea>
								<input type="file" id="fastafiles" name="fastafiles[]" multiple >
								<output id="fastafilesinfo"></output>

							</dd>

							<dt>
							<!-- Enter comma separated NCBI accessions --OR-- load FASTA files<br>
							</dd> -->
							</dl></p>

							<p><input type="submit" id="loadrefmap" value="Continue"/></p>
						</div>
					</div>

					<div class="panel panel-default" id="inputdatamenu">
						<div class="panel-heading"><label>Settings of MoDMap selected</label></div>
						<div class="panel-body">
							<p><div id="allInputDataMinimized">
								<a href="#" id="allInputDataShow">Show/Hide input data</a>
							</div></p>

							<div id="allInputData">
								<dl class="dl-horizontal">
									<dt>Length of k-mers:</dt>
									<dd>
										<select id="kmerslen">
											<option value="3">3 bp</option>
											<option value="4">4 bp</option>
											<option value="5">5 bp</option>
											<option value="6">6 bp</option>
											<option value="7">7 bp</option>
											<option value="8">8 bp</option>
											<option value="9">9 bp</option>
											<option value="10">10 bp</option>
											<option value="11">11 bp</option>
											<option value="12">12 bp</option>
										</select>
									</dd>
									<dt>Get Taxonomy:</dt>
									<dd>
										<select id="getTaxa">
											<option value="0">No (Fast)</option>
											<option value="1">Yes (Slow)</option>
										</select>
									</dd>
									<dt>Number of groups:</dt>
									<dd id="selectnumofsets">
										<select id="numofsets">
											<option value="0">Choose</option>
											<option value="1">1</option>
											<option value="2">2</option>
											<option value="3">3</option>
											<option value="4">4</option>
											<option value="5">5</option>
											<option value="6">6</option>
											<option value="7">7</option>
											<option value="8">8</option>
											<option value="9">9</option>
											<option value="10">10</option>
											<option value="11">11</option> 
											<option value="12">12</option>
											<option value="13">13</option>
											<option value="14">14</option>
											<option value="15">15</option> 
										</select>
									</dd>
								</dl>

								<div id="sets"></div>
							</div>

							<p><input type="submit" id="buildmap" value="Compute MoDMap" style="display:none;" /></p>
						</div>
					</div>

					<div class="panel panel-default" id="computemenu">
					<div class="panel-heading"><label>Computation Results</label></div>
					<div class="panel-body">
						<p id="stepstatus"></p>
						<div id="mainprogressbar" style="display:none;">
						  <div id="progress"></div>
						</div>
						<p id="file_contents"></p>
					</div>
					</div>

				</div>
			</div>
			<div id="menu3" class="tab-pane fade">
				<h3>Compute a distance</h3>
				<p>Compute approximated information distance (AID) between any pair of sequences, giving as input their NCBI accession numbers. AID distance has a value in [0,1]. If you do not know the accession number of a species, you can search directly in NCBI from <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Homo+sapiens" target="_blank">here</a>.</p>

				<div class="panel panel-default">
					<div class="panel-heading"><label>Settings</label></div>
					<div class="panel-body">
						<p>Samples you can try:
						<a href="#" id="aid1">[1] H.sapiens vs P.troglodytes</a> - 
						<a href="#" id="aid2">[2] H.sapiens vs H.neanderthalensis</a></p>
						Compute AID distance between (input NCBI accession numbers): 
						<ol>
							<li><input id="seq0" type="text" size="10" autofocus ></li>
							<li><input id="seq1" type="text" size="10" ></li>
						</ol>	
						<input type="submit" id="computedist" value="Compute AID distance" />
					</div>
				</div>

				<div class="panel panel-default"  id="aidinfomenu" style="display:none;">
					<div class="panel-heading"><label>Computation Results</label></div>
					<div class="panel-body"  id=aidinfo></div>
				</div>
			</div>
			<div id="menu4" class="tab-pane fade">
				<p>
				<dl class="dl-horizontal">
					<dt>Coded by:</dt>
					<dd>Rallis Karamichalis, 2016</dd>
					<dt>Github Code:</dt>
					<dd><a href="http://github.com/moleculardistancemaps/MoDMaps3D/" target="_blank">http://github.com/moleculardistancemaps/MoDMaps3D/</a></dd>
					<dt>HotKeys:</dt>
					
					<dd>
						<div id="hotkeysshowhide"><a href="#">Show/Hide full list</a></div>
						<div id="hotkeyslist">
						While on canvas:
						<ul>
						<li>A = Move camera Left</li>
						<li>D = Move camera Right</li>
						<li>W = Move camera Up</li>
						<li>S = Move camera Down</li>
						<li>E = Zoom camera In</li>
						<li>Q = Zoom camera Out</li>
						<li>Ctrl + LeftArrow = Rotate Left</li>
						<li>Ctrl + UpArrow = Rotate Up</li>
						<li>Ctrl + RightArrow = Rotate Right</li>
						<li>Ctrl + DownArrow = Rotate Down</li>
						<li>Left Mouse Click + Drag = Rotate to any direction</li>
						</ul>

						General Shortcuts:
						<ul>
						<li>Ctrl + F = Search field</li>
						<li>Ctrl + Alt + R = Reset Camera view, keeping rotation</li>
						<li>Ctrl + Alt + M = Toggle visibility of right menu</li>
						<li>Ctrl + Alt + N = Toggle visibility of left menu</li>
						</ul>
						</div>
					</dd>
					
				</dl>
				</p>	
			</div>
		</div>	
	</div>

</body>
</html>
